File: scripts/verify-hash-vectors.py
#!/usr/bin/env python3
"""
Verify Python implementation matches test vectors.
Run with: python scripts/verify-hash-vectors.py

This script validates that the Python canonical JSON and SHA256
implementations produce outputs that match the test vectors.

IMPORTANT: This is a RELEASE GATE requirement. This script must pass
before any release.
"""
import json
import hashlib
import sys
from pathlib import Path
from typing import Any, Dict, List, Union


def sort_keys_recursive(obj: Any) -> Any:
    """
    Recursively sort dictionary keys and remove null values.

    This matches the TypeScript canonicalize() function behavior:
    - Sort object keys lexicographically
    - Remove null values from objects
    - Preserve array order
    - Recursively process nested structures

    Args:
        obj: Any JSON-compatible value

    Returns:
        Normalized value with sorted keys and nulls removed
    """
    if isinstance(obj, dict):
        return {
            k: sort_keys_recursive(v)
            for k, v in sorted(obj.items())
            if v is not None
        }
    elif isinstance(obj, list):
        return [sort_keys_recursive(item) for item in obj]
    else:
        return obj


def canonical_json(obj: Any) -> str:
    """
    RFC 8785 (JCS) canonicalization with null removal.

    This produces a deterministic JSON string that matches the TypeScript
    implementation:
    - Sort keys lexicographically
    - Remove null values
    - No extra whitespace
    - UTF-8 encoding (ensure_ascii=False)

    Args:
        obj: Any JSON-compatible value

    Returns:
        Canonical JSON string
    """
    sorted_obj = sort_keys_recursive(obj)
    return json.dumps(sorted_obj, separators=(',', ':'), ensure_ascii=False)


def sha256_hash(data: str) -> str:
    """
    SHA256 hash of string, returns hex.

    Args:
        data: String to hash (will be UTF-8 encoded)

    Returns:
        Lowercase hex string of the SHA256 hash
    """
    return hashlib.sha256(data.encode('utf-8')).hexdigest()


def main() -> int:
    """
    Main verification function.

    Returns:
        0 if all vectors pass, 1 if any fail
    """
    fixtures_path = Path(__file__).parent.parent / 'fixtures' / 'hash-vectors.json'

    if not fixtures_path.exists():
        print(f'ERROR: Fixtures file not found at {fixtures_path}')
        print('Run "npx tsx scripts/generate-hash-vectors.ts" first.')
        return 1

    with open(fixtures_path, encoding='utf-8') as f:
        data = json.load(f)

    passed = 0
    failed = 0

    print('Verifying Python implementation against test vectors...\n')

    for vector in data['vectors']:
        name = vector['name']
        input_obj = vector['input']
        expected_canonical = vector['canonical']
        expected_hash = vector['sha256']

        actual_canonical = canonical_json(input_obj)
        actual_hash = sha256_hash(actual_canonical)

        canonical_match = actual_canonical == expected_canonical
        hash_match = actual_hash == expected_hash

        if canonical_match and hash_match:
            print(f'[PASS] {name}')
            passed += 1
        else:
            print(f'[FAIL] {name}')
            if not canonical_match:
                print(f'   Canonical mismatch:')
                print(f'     Expected: {expected_canonical}')
                print(f'     Actual:   {actual_canonical}')
            if not hash_match:
                print(f'   Hash mismatch:')
                print(f'     Expected: {expected_hash}')
                print(f'     Actual:   {actual_hash}')
            failed += 1

    print(f'\n{"=" * 50}')
    print(f'Results: {passed} passed, {failed} failed')

    if failed > 0:
        print('\nVERIFICATION FAILED - Release gate not passed!')
        return 1

    print('\nVERIFICATION PASSED - Python implementation is correct.')
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: scripts/verify-hash-vectors.ts
/**
 * Verify TypeScript implementation matches test vectors.
 * Run with: npx tsx scripts/verify-hash-vectors.ts
 *
 * This script validates that the TypeScript canonical JSON and SHA256
 * implementations produce outputs that match the test vectors.
 *
 * IMPORTANT: This is a RELEASE GATE requirement. This script must pass
 * before any release.
 */
import { readFileSync } from "fs";
import { join } from "path";
import { canonicalize, sha256StringHex } from "../packages/core/src/utils/index.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface Vector {
  name: string;
  description?: string;
  input: unknown;
  canonical: string;
  sha256: string;
}

interface VectorFile {
  description: string;
  version: string;
  generated: string;
  vectors: Vector[];
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const fixturesPath = join(import.meta.dirname ?? __dirname, "..", "fixtures", "hash-vectors.json");
  const vectors: VectorFile = JSON.parse(readFileSync(fixturesPath, "utf-8"));

  let passed = 0;
  let failed = 0;

  console.log("Verifying TypeScript implementation against test vectors...\n");

  for (const vector of vectors.vectors) {
    const actualCanonical = canonicalize(vector.input);
    const actualHash = await sha256StringHex(actualCanonical);

    const canonicalMatch = actualCanonical === vector.canonical;
    const hashMatch = actualHash === vector.sha256;

    if (canonicalMatch && hashMatch) {
      console.log(`[PASS] ${vector.name}`);
      passed++;
    } else {
      console.log(`[FAIL] ${vector.name}`);
      if (!canonicalMatch) {
        console.log(`   Canonical mismatch:`);
        console.log(`     Expected: ${vector.canonical}`);
        console.log(`     Actual:   ${actualCanonical}`);
      }
      if (!hashMatch) {
        console.log(`   Hash mismatch:`);
        console.log(`     Expected: ${vector.sha256}`);
        console.log(`     Actual:   ${actualHash}`);
      }
      failed++;
    }
  }

  console.log(`\n${"=".repeat(50)}`);
  console.log(`Results: ${passed} passed, ${failed} failed`);

  if (failed > 0) {
    console.log("\nVERIFICATION FAILED - Release gate not passed!");
    process.exit(1);
  }

  console.log("\nVERIFICATION PASSED - TypeScript implementation is correct.");
}

main().catch((err) => {
  console.error("Error verifying vectors:", err);
  process.exit(1);
});
